qiskit/quantum_info/synthesis/one_qubit_decompose.py

```diff
@@ -20,15 +20,17 @@
 import scipy.linalg as la
 
 from qiskit.circuit.quantumcircuit import QuantumCircuit
-from qiskit.circuit.library.standard_gates import (PhaseGate, U3Gate,
-                                                   U1Gate, RXGate, RYGate,
+from qiskit.circuit.library.standard_gates import (UGate, PhaseGate, U3Gate,
+                                                   U2Gate, U1Gate, RXGate, RYGate,
                                                    RZGate, RGate, SXGate)
 from qiskit.exceptions import QiskitError
 from qiskit.quantum_info.operators.predicates import is_unitary_matrix
 
 DEFAULT_ATOL = 1e-12
 
 ONE_QUBIT_EULER_BASIS_GATES = {
+    # 'U321' must precede 'U3' in the list
```

Comment:
Can you update the comment to explain why this is? If this is for the `Optimize1qGatesDecomposer`, it would be good to add a comment there and open an issue to find a better metric than depth.

Reply:
This is the only case in the dict where one basis set is a subset of the other. If the basis passed to the tranpiler has ['u3', 'u2', 'u1'], you'd want the decomposer to use circuit_u321 first before circuit_u3. If the order is reversed it fails 2 tests that were added. Since it's a dict, this may need further thought.

Reply:
On further thought, perhaps use an OrderedDict and an appropriate general comment about subsets needing to succeed supersets in the order.

Reply:
An alternative would be to have Optimize1qGatesDecomposition as it iterates the dict to check if any previously added basis is a superset of the current one and skip the current one if it's true, or if a previous one is a subset of this one, remove the previous one. Since the dict is small, this should not be a performance issue, and would allow any ordering of the dict.

Reply:
Went ahead and did what I suggested above in O1GD. Done in https://github.com/Qiskit/qiskit-terra/pull/5554/commits/b61b70954cbdd4b8719437057c9c6642441ac286 and https://github.com/Qiskit/qiskit-terra/pull/5554/commits/aeed31e49f7fcac65da3b0b974408df6ba20bf3f